File: minecraft.js
const readline = require('readline');

// Константы блоков
const AIR = ' ';
const DIRT = '.';
const STONE = '#';
const BEDROCK = 'X';
const PLAYER = 'P';

class World {
    constructor(size) {
        this.size = size;
        this.blocks = new Array(size).fill(AIR);
        this.generateTerrain();
    }

    generateTerrain() {
        for (let i = 0; i < this.size; i++) {
            if (i === 0 || i === this.size - 1) this.blocks[i] = BEDROCK;
            else if (i < 4) this.blocks[i] = STONE;
            else if (i < 8) this.blocks[i] = DIRT;
            else this.blocks[i] = AIR;
        }
    }

    render(playerPos, inventory) {
        // Очистка экрана консоли
        console.clear();

        console.log("=== 1D MINECRAFT НА JAVASCRIPT (Node.js) ===");
        console.log();

        // Отрисовка мира
        let displayLine = "";
        for (let i = 0; i < this.size; i++) {
            if (i === playerPos) {
                displayLine += PLAYER;
            } else {
                displayLine += this.blocks[i];
            }
        }
        console.log(displayLine);

        // Интерфейс
        const invStr = inventory === AIR ? "Пусто" : inventory;
        console.log(`\n[Инвентарь]: ${invStr}`);
        console.log("[Управление]: A (Влево) | D (Вправо) | Q (Копать под собой) | E (Строить под собой) | X (Выход)");
        process.stdout.write("Ввод: ");
    }
}

function main() {
    const worldSize = 40;
    const world = new World(worldSize);

    let playerPos = 12;
    let inventory = AIR;

    // Настройка интерфейса чтения ввода из консоли
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    const gameLoop = () => {
        // Гравитация
        while (playerPos > 0 && world.blocks[playerPos - 1] === AIR) {
            playerPos--;
        }

        world.render(playerPos, inventory);
    };

    // Запускаем первый кадр
    gameLoop();

    rl.on('line', (line) => {
        const action = line.trim().toLowerCase();

        if (action === 'x') {
            console.log("Игра завершена!");
            rl.close();
            process.exit(0);
        }

        switch (action) {
            case 'a': // Влево
                if (playerPos > 0 && world.blocks[playerPos - 1] === AIR) {
                    playerPos--;
                }
                break;
            case 'd': // Вправо
                if (playerPos < worldSize - 1 && world.blocks[playerPos + 1] === AIR) {
                    playerPos++;
                }
                break;
            case 'q': // Копать под собой
                if (playerPos > 0) {
                    const target = world.blocks[playerPos - 1];
                    if (target !== AIR && target !== BEDROCK && inventory === AIR) {
                        inventory = target;
                        world.blocks[playerPos - 1] = AIR;
                    }
                }
                break;
            case 'e': // Строить под собой
                if (inventory !== AIR && playerPos > 1 && world.blocks[playerPos - 1] === AIR) {
                    world.blocks[playerPos - 1] = inventory;
                    inventory = AIR;
                    playerPos++;
                }
                break;
        }

        // Обновляем состояние после хода
        gameLoop();
    });
}

main();

File: minecraft.py
import os
import platform

# Константы для отображения блоков
AIR = ' '
DIRT = '.'
STONE = '#'
BEDROCK = 'X'
PLAYER = 'P'

class World:
    def __init__(self, size):
        self.size = size
        self.blocks = [AIR] * size
        self.generate_terrain()

    def generate_terrain(self):
        """Генерация слоев одномерного ландшафта"""
        for i in range(self.size):
            if i == 0 or i == self.size - 1:
                self.blocks[i] = BEDROCK
            elif i < 4:
                self.blocks[i] = STONE
            elif i < 8:
                self.blocks[i] = DIRT
            else:
                self.blocks[i] = AIR

    def render(self, player_pos, inventory):
        """Очистка экрана и отрисовка игрового поля"""
        # Кроссплатформенная очистка консоли
        if platform.system() == "Windows":
            os.system('cls')
        else:
            os.system('clear')

        print("=== 1D MINECRAFT НА PYTHON ===")
        print()

        # Отрисовка мира
        display_world = []
        for i in range(self.size):
            if i == player_pos:
                display_world.append(PLAYER)
            else:
                display_world.append(self.blocks[i])
        
        print("".join(display_world))

        # Вывод интерфейса
        inv_str = "Пусто" if inventory == AIR else inventory
        print(f"\n[Инвентарь]: {inv_str}")
        print("[Управление]: A (Влево) | D (Вправо) | Q (Копать под собой) | E (Строить под собой) | X (Выход)")
        
def main():
    world_size = 40
    world = World(world_size)
    
    player_pos = 12
    inventory = AIR
    running = True

    while running:
        # Физика гравитации: персонаж падает, пока внизу воздух
        while player_pos > 0 and world.blocks[player_pos - 1] == AIR:
            player_pos -= 1

        world.render(player_pos, inventory)
        
        # Получение ввода пользователя
        try:
            user_input = input("Ввод: ").strip().lower()
        except (KeyboardInterrupt, EOFError):
            break

        if not user_input:
            continue

        action = user_input[0]

        if action == 'x':
            running = False

        elif action == 'a':  # Движение влево
            if player_pos > 0 and world.blocks[player_pos - 1] == AIR:
                player_pos -= 1

        elif action == 'd':  # Движение вправо
            if player_pos < world_size - 1 and world.blocks[player_pos + 1] == AIR:
                player_pos += 1

        elif action == 'q':  # Разрушение блока под собой
            if player_pos > 0:
                target = world.blocks[player_pos - 1]
                if target != AIR and target != BEDROCK and inventory == AIR:
                    inventory = target
                    world.blocks[player_pos - 1] = AIR

        elif action == 'e':  # Строительство блока под собой
            if inventory != AIR and player_pos > 1 and world.blocks[player_pos - 1] == AIR:
                world.blocks[player_pos - 1] = inventory
                inventory = AIR
                player_pos += 1  # Смещаем игрока вверх на построенный блок

    print("Игра завершена!")

if __name__ == "__main__":
    main()
